### PR TITLE
Gadgets, Clients, etc.

### PR DESCRIPTION
**Pull request by mbechler**
Sorry for the size of this and some unnecessary formatting changes.

A few more gadget RCE chains:
- Hibernate5/Hibernate4
- MyFaces
- C3P0
- net.sf.json-lib
(beanutils you already got ;))

Non RCE Chains:
- FileUpload (some potentially dangerous file ops still possible)
- JavaSE (opening an JRMP listener/client, nice for bypassing application level filters)

In case you are interested I also just published my tooling that found these at
https://github.com/mbechler/serianalyzer (it's a static bytecode analyzer that traces native calls reachable by deserialization, many FPs of course)

And some explotitation tooling:
- JRMP client/server
- JSF ViewState
- Jenkins CLI client (+ my POCs for CVE-2016-0788)
- A JBoss/WildFly RMI client

### CODE REVIEW

File: src/main/java/ysoserial/payloads/Spring2.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            Type.class,
+            Templates.class);
+
+        final Object typeProviderProxy = Gadgets.createMemoitizedProxy(
+            Gadgets.createMap("getType", typeTemplatesProxy),
+            forName("org.springframework.core.SerializableTypeWrapper$TypeProvider"));
+
+        ReflectionFactory rf = ReflectionFactory.getReflectionFactory();
+        Constructor<?> sc = rf.newConstructorForSerialization(
+            Class.forName("org.springframework.core.SerializableTypeWrapper$MethodInvokeTypeProvider"),
```
Comment (frohoff) [inline src/main/java/ysoserial/payloads/Spring2.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Class.forName' due to existing static import 'java.lang.Class.forName'](https://www.codacy.com/app/frohoff/ysoserial/file/2167879616/issues/source?bid=3161887&fileBranchId=3161889#l65)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+        FutureResult<Channel> chResult = new FutureResult<Channel>(context.getExecutor());
+        ch.open("jmx", chResult, options);
+
+        IoFuture<Channel> cFuture = chResult.getIoFuture();
+        Status s2 = cFuture.await();
+        if ( s2 == Status.FAILED ) {
+            System.err.println("Cannot connect");
+            if ( cFuture.getException() != null ) {
+                cFuture.getException().printStackTrace(System.err);
+                System.exit(-1);
```
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [System.exit() should not be used in J2EE/JEE apps](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l294)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+    /**
+     * @author mbechler
+     *
+     */
+    private static final class ConsoleLogHandler extends Handler {
+
+        /**
+         * 
+         */
+        public ConsoleLogHandler () {}
```
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [Avoid unnecessary constructors - the compiler will generate these for you](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l382)

File: src/main/java/ysoserial/payloads/Spring2.java (hunk 1 above)
Comment (frohoff) [inline src/main/java/ysoserial/payloads/Spring2.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Class.forName' due to existing static import 'java.lang.Class.forName'](https://www.codacy.com/app/frohoff/ysoserial/file/2167879616/issues/source?bid=3161887&fileBranchId=3161889#l65)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+ * 
+ * @author mbechler
+ *
+ */
+@SuppressWarnings ( {
+    "nls", "javadoc", "rawtypes"
+} )
+public class JBoss {
+
+    public static void main ( String[] args ) {
```
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [The method main() has an NPath complexity of 252](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l91)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java (hunk 2 above)
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [System.exit() should not be used in J2EE/JEE apps](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l294)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+
+        /**
+         * @param context
+         */
+        public ConnectionHandlerContextImpl ( ConnectionProviderContextImpl context ) {
+            this.context = context;
+        }
+
+
+        public void remoteClosed () {}
```
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l418)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java (hunk 3 above)
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [Avoid unnecessary constructors - the compiler will generate these for you](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l382)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+
+
+        @Override
+        public void publish ( LogRecord record ) {
+            System.err.println(record.getMessage());
+        }
+
+
+        @Override
+        public void flush () {
```
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l392)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java (hunk 4 above)
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [The method main() has an NPath complexity of 252](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l91)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+
+                for ( Callback cb : callbacks ) {
+
+                    if ( cb instanceof NameCallback ) {
+                        ( (NameCallback) cb ).setName(username);
+                    }
+                    else if ( cb instanceof PasswordCallback ) {
+                        ( (PasswordCallback) cb ).setPassword(password != null ? password.toCharArray() : new char[0]);
+                    }
+                    else if ( cb instanceof RealmCallback ) {
```
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [Avoid empty if statements](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l245)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java (hunk 5 above)
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l418)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+
+
+        @Override
+        public void flush () {
+
+        }
+
+
+        @Override
+        public void close () throws SecurityException {}
```
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l398)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java (hunk 6 above)
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l392)

File: src/main/java/ysoserial/exploit/JSF.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            System.exit(-1);
+            return;
+        }
+
+        try {
+            URL u = new URL(args[ 0 ]);
+
+            URLConnection c = u.openConnection();
+            if ( ! ( c instanceof HttpURLConnection ) ) {
+                throw new Exception("Not a HTTP url"); //$NON-NLS-1$
```
Comment (frohoff) [inline src/main/java/ysoserial/exploit/JSF.java]: Codacy Issue found: [Avoid throwing raw exception types.](https://www.codacy.com/app/frohoff/ysoserial/file/2167879596/issues/source?bid=3161887&fileBranchId=3161889#l58)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java (hunk 7 above)
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [Avoid empty if statements](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l245)

File: src/main/java/ysoserial/exploit/JSF.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+public class JSF {
+
+    /**
+     * @param args
+     */
+    public static void main ( String[] args ) {
+
+        if ( args.length < 3 ) {
+            System.err.println(JSF.class.getName() + " <view_url> <payload_type> <payload_arg>");
+            System.exit(-1);
```
Comment (frohoff) [inline src/main/java/ysoserial/exploit/JSF.java]: Codacy Issue found: [System.exit() should not be used in J2EE/JEE apps](https://www.codacy.com/app/frohoff/ysoserial/file/2167879596/issues/source?bid=3161887&fileBranchId=3161889#l31)

File: src/main/jboss/ysoserial/exploit/jboss/JBoss.java (hunk 8 above)
Comment (frohoff) [inline src/main/jboss/ysoserial/exploit/jboss/JBoss.java]: Codacy Issue found: [Document empty method body](https://www.codacy.com/app/frohoff/ysoserial/file/2167879625/issues/source?bid=3161887&fileBranchId=3161889#l398)

File: src/main/java/ysoserial/exploit/JSF.java (hunk 9 above)
Comment (frohoff) [inline src/main/java/ysoserial/exploit/JSF.java]: Codacy Issue found: [Avoid throwing raw exception types.](https://www.codacy.com/app/frohoff/ysoserial/file/2167879596/issues/source?bid=3161887&fileBranchId=3161889#l58)

File: src/main/java/ysoserial/exploit/JSF.java (hunk 10 above)
Comment (frohoff) [inline src/main/java/ysoserial/exploit/JSF.java]: Codacy Issue found: [System.exit() should not be used in J2EE/JEE apps](https://www.codacy.com/app/frohoff/ysoserial/file/2167879596/issues/source?bid=3161887&fileBranchId=3161889#l31)

**Comment by frohoff** [comment]
This is some great stuff and the analysis tool also looks fantastic.

As you said, it's a bit of a monster PR, so I was hoping we could tidy it up and trim/split it a bit, especially with some possibly non-trivial refactors on the horizon. Here's some immediate thoughts:

* **JDK Compatibility**: There seem to be some dependencies on JDK8 classes (namely `java.util.Base64`). Can these be removed and make sure it builds against at least JDK7 (and JDK8)?
* **Git History**: There's 3 merge commits in here. Perhaps this can call be rebased into a single commit against master?
* **License/Copyright**: I'm no software license expert, but I don't think the "All rights reserved" (non-)licensing works inside an MIT licensed project. Can these be removed? I think Copyright (and definitely credit comments) are fine. Also see #15 
* **Unnecessary Changes**: As you mentioned, there's a lot of whitespace/formatting changes to existing files, as well as many `@SupressWarnings` argument and `$NON-NLS-1$` comment additions. Can these be removed/reverted to simplify the PR/merge? If some of these really do make sense, they should probably just be added to the whole codebase in a separate commit.
* **Tests**: These things are certainly not easy to test, but with some larger refactors possibly coming down the pipe, it would be nice to know we weren't breaking gadgets/chains. Is it possible to build a test harness for some of these JNDI remote-load payloads and perhaps the commons-fileupload one? 
* **Maven Profiles**: I'm not sure if/why some of these maven profiles are necessary. Can you elaborate on the goals with this and/or if any of it could be simplified? For example, can the jenkins remoting jar be pulled down normally from [this repo](http://repo.jenkins-ci.org/public/org/jenkins-ci/main/remoting/2.55/)? Is there a way to make it play nicely with IDEs? #10 may be relevant to this.

Again, this is some great work and I'm hoping we can get it pulled in without too much trouble. If you're short on time we can resubmit the PR against a different branch and I (or other contributors) can take a crack at getting pieces integrated.

Either way, thanks for the contribution, and please continue looking for more gadgets/chains and educating people about unsafe deserialization.

**Comment by mbechler** [comment]
Superseded by #20

**Pull request closed without merging**